### PR TITLE
fix(#1281): HexColorField accessibility updates

### LINKED ISSUE #1281: HexColorField accessibility review
# 📝 Feedback

1. The Popover examples demonstrate a need in the Color API for being able to determine the luminosity of a Color, the contrast ratio between one Color and another, and whether a particular color has sufficient contrast with another color to comply with the WCAG 2.1 [1.4.3 Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) or [1.4.6 Contrast (Enhanced)](https://www.w3.org/TR/WCAG21/#contrast-enhanced) success criteria. This would allow a developer to change the foreground color of text to have sufficient contrast with a background color.

2. While I understand the choice to make the HexColorField a `role="spinbutton"`, like the Hex color inputs in Google Chrome's Web Developer tools, I'm not sure that incrementing between 0 and 16,777,215 is particularly useful. Stepper functionality probably makes more sense when the input format is `rgb`, `rgba`, `hsl`, `hsla`, where its easier to distinguish between component parts of the color. There is a good example of this distinction in the inputs on https://contrast-ratio.com/, where, if the user inputs a hex value, using the ArrowUp or ArrowDown keys does not change the value.

3. With VoiceOver on iOS, `role="spinbutton"` makes it much harder to bring up the onscreen keyboard. Rather than a simple double tap, I have to double tap and hold to open the on screen keyboard. This is particularly problematic in the Popover stories, like https://reactspectrum.blob.core.windows.net/reactspectrum/32a59e198acd249536c92347767eedfcd3470590/storybook-17/index.html?path=/story/hexcolorfield--as-a-popover, because double tapping the HexColorField focused with the VoiceOver cursor within the Popover closes the popover. One can open the dialog and read the the text input, but when the user double taps to edit the text using the on-screen keyboard, the popover closes, because the field loses focus when the popover repositions.

4. We should use `Content` instead of `View` as the content slot for the `Dialog` in the Storybook story.

5. The `validationState="valid"` check image is not communicated in any way to assistive technology.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| @react-spectrum/color   |  3.0.0-alpha.1
| Browser          | All
| Operating System | All

## 🧢 Your Company/Team

Adobe/Accessibility

### PR DESCRIPTION
**Pull request by majornista**
Closes #1281 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue 1281](https://github.com/adobe/react-spectrum/issues/1281).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Unit tests have been updated
2. For http://localhost:9003/?path=/story/hexcolorfield--as-a-popover: Verify that Button text color updates to provide sufficient contrast against background color when HexColorField value updates.
3. For http://localhost:9003/?path=/story/hexcolorfield--default: Using VoiceOver on iOS, verify that double tapping on the HexColorField expands on-screen keyboard without having to double tap and hold.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-stately/color/src/Color.ts [hunk 1]
```diff
@@ -91,6 +91,65 @@ export class Color {
...
+    let comparisonColorLum = comparisonColor.luminance();
+
+    if (colorLum > comparisonColorLum) {
+      return (colorLum + 0.05) / (comparisonColorLum + 0.05);
+    }
+
+    return (comparisonColorLum + 0.05) / (colorLum + 0.05);
+  }
+
+  wcagComplianceLevel(comparisonColor: Color): string {
```
Comment (snowystinger) [inline packages/@react-stately/color/src/Color.ts:135]: this seems more like a dev tool? do we want it in our color code? or should we just provide some code in documentation that knows how to handle our color objects?

Comment (majornista) [reply packages/@react-stately/color/src/Color.ts:135]: Fair enough.

File: packages/@react-stately/color/src/Color.ts [hunk 2]
```diff
@@ -91,6 +91,56 @@ export class Color {
...
+
+  isDark() {
+    var yiq = (this.getChannelValue('red') * 299 + this.getChannelValue('green') * 587 + this.getChannelValue('blue') * 114) / 1000;
+    return yiq < 128;
+  }
+
+  isLight() {
+    return !this.isDark();
+  }
+
```
Comment (devongovett) [inline packages/@react-stately/color/src/Color.ts:143]: Looks like these are only used in the example story? Not sure if we'll need them in any of the real components. If we can we should try to keep the API surface area/bundle size minimal...

**Comment by devongovett** [comment]
Moved to #1495

**Pull request closed without merging**